### ISSUE TITLE
Feature request - collapse empty directories

### ISSUE THREAD

**Issue by ambihelical** [Alexander-Miller/treemacs#31]
It would be great if treemacs could collapse empty directories.
Github does this--see for example the code tab at Alexander-Miller/dotfiles, specifically the .themes directory.  
Another example is the emacs package dired-collapse, which adds this functionality to dired-mode.

**Comment by Alexander-Miller** [comment]
It's a good option to have, but I am very uncomfortable with the performance hit. Just adding a recursive check for the number of files in every dir when building a tree for 1000 dirs doubles the runtime for me.

Based on my own testing (opening /bin with 3000 files a couple of times) treemacs is currently the fastest project explorer around, and I'd like to keep it that way. Therefore I'd much prefer an external script that can be run in parallel to treemacs' own tree creation, the way it's already done with git. Unless you've a better idea or know where I should look I'll see what I can do with python or bash.

**Comment by ambihelical** [comment]
I'm not sure what the code was doing that you mentioned, but if it is counting all files in any directory then it is doing too much.  If you recurse into a directory and find any files or more than one directory, then you can't collapse any further.   Usually the cases where you can collapse are fairly rare, so it should not be a major hit. I guess it could be an option or a toggle if the impact is otherwise.   Thanks for considering it in any case.

**Comment by Alexander-Miller** [comment]
I just added these lines:
```
(when (= 1 (length (f-entries node nil t)))
    (setq ff (1+ ff)))
```
All the code does is look at the amount of entries, that is both dirs and files, inside a directory before inserting a button for it, and increments a counter if that number was 1. Recursion doesn't matter since I ran it on 1000 empty directories. 

The problem is that AFAICT there's nothing I can do to filter directories. At the very least I need to check for every single one if its content is nothing but another directory. That's potentially a lot of calls into the file system. Luckily it's easy to parallelize these checks.

At any rate, this will go on my todo list, but for now I'd close some previous issues and concentrate on cleanup and tests. Dunno if you noticed, but treemacs, was broken for a few days due to issues that only appeared when it's installed from melpa.

**Comment by Alexander-Miller** [comment]
I'm mostly done with the feature now, only cleanup and writing documentation should be left. I've pushed what I have [to a feature branch](https://github.com/Alexander-Miller/treemacs/tree/collapse), be great if you could look at it and make sure it's well behaved for your use-cases.

You'll need to set `treemacs-collapse-dirs` to t, though later it'll be a number to limit how many directories can be collapsed at once.

**Comment by ambihelical** [comment]
It seems to work quite well, but I only briefly tried it.  I'll try it on some java projects I have at work.

**Comment by Alexander-Miller** [comment]
Change is on master now, give it a go and we can close up here.

**Comment by ambihelical** [comment]
I haven't had any problems with it (pre-master), will try it out on master this coming week.

**Comment by ambihelical** [comment]
I haven't had any problems on a fairly large code base. I've been using 3 for max dirs to collapse.  The only issue I've had is the non-resizable window, but that is not related to this code I think and is a problem as well with deep directory structures. I added #41 for it.

**Comment by Alexander-Miller** [comment]
>I haven't had any problems on a fairly large code base.

You must've tried recently then. I needed to amend the melpa recipe to make sure the python file is included in the build.

> and is a problem as well with deep directory structures.

Resizing aside, there's a few more things you can do for deep nesting:
 * collapse more dirs
 * use a smaller font in all the treemacs faces
 * decrease `treemacs-indentation`

**Comment by ambihelical** [comment]
Yes, fairly recently, but using master.  

I really just need to make the window bigger sometimes, my font is as small as I can manage.  I didn't know about treemacs-toggle-fixed-width though.

**Comment by Alexander-Miller** [comment]
I think we can close up here now.

**Comment by minad** [comment]
@Alexander-Miller I've seen you are shelling out to python for certain async features like collapsing nodes. Would something like that also be possible with the integrated threading system of emacs? I've heard the newer versions have something like that.

**Comment by Alexander-Miller** [comment]
Not the same thing. Threads in elisp are cooperative. Basically you can have multiple threads, but only 1 will ever be running. There's no real parallelism, it just makes writing IO heavy code a lot easier since a thread that's waiting on IO will automatically yield.

I start python processes specifically so I can run elisp and python in parallel. Python searches the file system for nodes to collapse while at the same time elisp is busy building and displaying the content of the node you've just expanded.

**Comment by minad** [comment]
Meh, I thought there are real native worker threads :( I had some issues with sentinel errors when closing emacs which are due to those python processes I think, so this might have made stuff more robust.

**Comment by Alexander-Miller** [comment]
It's very unlikely that those are treemacs processes. They are UI helpers that finish within tens of ms and treemacs explicitly waits for them to be done.

What do your errors look like?

**Comment by minad** [comment]
I have no idea yet on how to debug those since they only occur when I am closing emacs. I guess it is more probable then that they are lsp errors due to problems with the server shutdown.

**Comment by Alexander-Miller** [comment]
If it is an error and not just a warning and debug-on-error is enabled then emacs won't shut down, and show you a stack trace instead.

**Comment by minad** [comment]
It is clearly an error. It says "error in process sentinel...". And unfortunately it does not stop in the debugger with debug-on-error=t. 

As I said I am pretty green concerning emacs lisp debugging. I looked around if it is possible to log the message buffer to some log file so see what is going on, but didn't find a way to do it. I would also like an option to print stack traces in the case of arbitrary errors, even the ones which were caught. Is there a way to do that? I've also seen there is bughunter, is that useful? Or do you have some other hints on how to approach these debugging issues I am having?

**Comment by Alexander-Miller** [comment]
Some stuff (like functions on a timer) basically runs with demoted errors and won't throw up a debugger when something goes wrong. I wouldn't know how to get a stack trace out of your situation either, but writing your log when quitting is quite easy:

```
(defun write-log-on-quit ()
  (with-current-buffer (get-buffer "*Messages*")
    (f-write (buffer-string) 'utf-8 "~/Documents/LOG")))
(add-hook 'kill-emacs-hook #'write-log-on-quit)
```
Requires f.el to be loaded, which will be if you've also loaded treemacs.

>  would also like an option to print stack traces in the case of arbitrary errors, even the ones which were caught

I don't think that is possible, unless you directly rewrite the catch block yourself.

Never heard of bug hunter before, but I don't think it's the way to go, the errors you see aren't part of a file and it sounds like they arent't even deterministic.

**Comment by minad** [comment]
> Some stuff (like functions on a timer) basically runs with demoted errors and won't throw up a debugger when something goes wrong. I wouldn't know how to get a stack trace out of your situation either, but writing your log when quitting is quite easy:

Thank you very much! I will try that!

> I don't think that is possible, unless you directly rewrite the catch block yourself.

Maybe it is possible to advice catch somehow. I would wonder if such a thing is not possible in a lisp ;)

**Comment by minad** [comment]
:( without debug-on-error it shows the sentinel error and with debug-on-error nothing shows up. Some undesired side effect of debug-on-error. This explains also why emacs is still quitting as normal.

I tried:

~~~ elisp
(defun write-log-on-quit ()
  (let ((b (get-buffer "*Messages*")))
    (if b
      (with-current-buffer b
        (f-write (buffer-string) 'utf-8 "~/.emacs.d/messages"))))
  (let ((b (get-buffer "*Backtrace*")))
    (if b
      (with-current-buffer b
        (f-write (buffer-string) 'utf-8 "~/.emacs.d/backtrace")))))
(add-hook 'kill-emacs-hook #'write-log-on-quit)
~~~

**Comment by Alexander-Miller** [comment]
By default a process sentinel will show a stack trace when debug-on-error is enabled, so whatever you see is likely running inside `with-demoted-errors`. In that case the messages logging should hopefully show us where it's coming from.

> Maybe it is possible to advice catch somehow.

You can do a lot with elisp, but rewriting C primitives is not a line that can be crossed.
